ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/OptimizedMerkleTrieLoader.java

```diff
@@ -0,0 +1,136 @@
+    CompletableFuture.runAsync(
+        () -> {
+          final StoredMerklePatriciaTrie<Bytes, Bytes> accountTrie =
+              new StoredMerklePatriciaTrie<>(
+                  (location, hash) -> {
+                    Optional<Bytes> node =
+                        worldStateStorage.getAccountStateTrieNode(location, hash);
+                    node.ifPresent(bytes -> accountsNodes.put(Hash.hash(bytes), bytes));
+                    return node;
+                  },
+                  worldStateRootHash,
+                  Function.identity(),
+                  Function.identity());
+          accountTrie.get(Hash.hash(account));
+        });
+  }
+
+  public void preLoadStorage(
+      final BonsaiWorldStateKeyValueStorage worldStateStorage,
+      final Address account,
+      final Hash slotHash) {
+    CompletableFuture.runAsync(
+        () -> {
+          final Hash accountHash = Hash.hash(account);
+          worldStateStorage
+              .getStateTrieNode(Bytes.concatenate(accountHash, Bytes.EMPTY))
+              .ifPresent(
+                  storageRoot -> {
+                    final StoredMerklePatriciaTrie<Bytes, Bytes> storageTrie =
+                        new StoredMerklePatriciaTrie<>(
+                            (location, hash) -> {
+                              Optional<Bytes> node =
+                                  worldStateStorage.getAccountStorageTrieNode(
+                                      accountHash, location, hash);
+                              node.ifPresent(bytes -> storageNodes.put(Hash.hash(bytes), bytes));
+                              return node;
+                            },
+                            Hash.hash(storageRoot),
+                            Function.identity(),
+                            Function.identity());
+                    storageTrie.get(slotHash);
+                  });
+        });
+  }
+
+  public Optional<Bytes> getAccountStateTrieNode(
+      final BonsaiWorldStateKeyValueStorage worldStateKeyValueStorage,
+      final Bytes location,
+      final Bytes32 nodeHash) {
+    if (nodeHash.equals(MerklePatriciaTrie.EMPTY_TRIE_NODE_HASH)) {
+      return Optional.of(MerklePatriciaTrie.EMPTY_TRIE_NODE);
+    } else {
+      Optional<Bytes> node = Optional.ofNullable(accountsNodes.getIfPresent(nodeHash));
+      if (node.isPresent()) {
+        return node;
+      } else {
+        return worldStateKeyValueStorage.getAccountStateTrieNode(location, nodeHash);
+      }
+    }
+  }
+
+  public Optional<Bytes> getAccountStorageTrieNode(
+      final BonsaiWorldStateKeyValueStorage worldStateKeyValueStorage,
+      final Hash accountHash,
+      final Bytes location,
+      final Bytes32 nodeHash) {
+    if (nodeHash.equals(MerklePatriciaTrie.EMPTY_TRIE_NODE_HASH)) {
+      return Optional.of(MerklePatriciaTrie.EMPTY_TRIE_NODE);
+    } else {
+      Optional<Bytes> node = Optional.ofNullable(storageNodes.getIfPresent(nodeHash));
+      if (node.isPresent()) {
```

Comment:
These code feels like it could be more functional.

Reply:
done